### PR TITLE
fix markdown editor italic command

### PR DESCRIPTION
**Pull request by HanyaElhashemy**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] ~~Server: I added multiple integration tests (Spring) related to the features~~
- [x] ~~Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)~~
- [x] ~~Server: I implemented the changes with a good performance and prevented too many database calls~~
- [x] ~~Server: I documented the Java code using JavaDoc style.~~
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] ~~Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)~~
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] ~~Client: I translated all the newly inserted strings into German and English~~

### Motivation and Context

resolved issue #934 

### Description

See pull request #904 description.
In addition some code refactoring to minimize duplicating the code: in the classes bold.command and italic.command only modified selected string by either adding (**) in bold.command or (*) in italic.command.
To remove the markdown we no longer check if text consists the markdown since it could consist it in the middle of selected text not at the beginning and in this case slicing the string would be wrong, instead we check if the border of the selected text is the markdown

### Steps for Testing

1.navigate to course administration
2.edit an existing exercise
3.write text in one of the editors for example in problem statement text editor
4.select text that start or/and ends with white-space
5.click on the"I" (Italic) button
you should see the markdowns only around actual text
6.click preview to see your edited text

### Screenshots

image
image
image

### CODE REVIEW

File: src/main/webapp/app/markdown-editor/commands/command.ts [hunk 1]
```diff
@@ -79,4 +79,31 @@ export abstract class Command {
     }
 
     abstract execute(input?: string): void;
+
+    protected ignoreWhiteSpace(text: string) {
+        const textLength = text.length;
+        let startSpace = '';
+        let endSpace = '';
+        let startIndx = 0;
```
Comment (lovaiible) [inline src/main/webapp/app/markdown-editor/commands/command.ts]:
```suggestion
        let startIndex = 0;
```

File: src/main/webapp/app/markdown-editor/commands/command.ts [hunk 2]
```diff
@@ -79,4 +79,31 @@ export abstract class Command {
...
+        let endIndx = 0;
```
Comment (lovaiible) [inline src/main/webapp/app/markdown-editor/commands/command.ts]:
```suggestion
        let endIndex = 0;
```

File: src/main/webapp/app/markdown-editor/commands/bold.command.ts [hunk 3]
```diff
@@ -15,34 +15,15 @@ export class BoldCommand extends Command {
...
-            for (let j = selectedTextLength; j >= 0; j--) {
-                if (selectedText.charAt(j - 1) === ' ') {
-                    endSpace = endSpace + ' ';
-                } else {
-                    endIndx = j;
-                    break;
-                }
-            }
-            const newTxt = selectedText.slice(startIndx, endIndx);
+            const refinedString = this.ignoreWhiteSpace(selectedText);
```
Comment (lovaiible) [inline src/main/webapp/app/markdown-editor/commands/bold.command.ts]: Except for the command characters `**`, this is has the same code as the in the `italic.command.ts` file. Maybe just move the creation of the `nexTxt` inside a new method that returns the final text to add with the command characters based on the caller.

Comment (HanyaElhashemy) [reply src/main/webapp/app/markdown-editor/commands/bold.command.ts]: Thank you for your constructive feedback, I modified my code accordingly !

File: src/main/webapp/app/markdown-editor/commands/bold.command.ts [hunk 4]
```diff
@@ -15,36 +15,14 @@ export class BoldCommand extends Command {
...
         const selectedText = this.getSelectedText();
         let textToAdd = '';
-        let startSpace = '';
-        let endSpace = '';
-        let startIndx = 0;
-        let endIndx = 0;
-        const selectedTextLength = selectedText.length;
 
-        if (selectedText.includes('**')) {
+        if (selectedText.charAt(0) === '*' && selectedText.charAt(selectedText.length - 1) === '*') {
```
Comment (thilo-behnke) [inline src/main/webapp/app/markdown-editor/commands/bold.command.ts]: Good fix!
However I think it should be "**" instead of "*" here, right?

Otherwise it is the same as in https://github.com/ls1intum/Artemis/pull/937/files#diff-a6bae12e2a7d3f19137d1bb31cab1fb9R19.

Comment (thilo-behnke) [inline src/main/webapp/app/markdown-editor/commands/bold.command.ts]: Good fix!
However I think it should be "**" instead of "*" here, right?

Otherwise it is the same as in https://github.com/ls1intum/Artemis/pull/937/files#diff-a6bae12e2a7d3f19137d1bb31cab1fb9R19.

File: src/main/webapp/app/markdown-editor/commands/bold.command.ts [hunk 5]
```diff
@@ -15,36 +15,14 @@ export class BoldCommand extends Command {
...
-                    endSpace = endSpace + ' ';
-                } else {
-                    endIndx = j;
-                    break;
-                }
-            }
-            const newTxt = selectedText.slice(startIndx, endIndx);
-            textToAdd = `**${newTxt}**`;
-            this.insertText(startSpace + textToAdd + endSpace);
+            const refinedString = this.ignoreWhiteSpace(selectedText);
```
Comment (thilo-behnke) [inline src/main/webapp/app/markdown-editor/commands/bold.command.ts]: Not a good method name - what is a `refinedString`?

File: src/main/webapp/app/markdown-editor/commands/bold.command.ts [hunk 6]
```diff
@@ -15,36 +15,14 @@ export class BoldCommand extends Command {
...
+            textToAdd = `**${refinedString}**`;
+            this.addRefinedTxt(textToAdd);
```
Comment (thilo-behnke) [inline src/main/webapp/app/markdown-editor/commands/bold.command.ts]: Try not to use unnecessary shortenings as e.g. txt, there is only one letter difference and it makes the reader wonder (is this related to a windows txt file? :wink: )

File: src/main/webapp/app/markdown-editor/commands/command.ts [hunk 7]
```diff
@@ -79,4 +81,35 @@ export abstract class Command {
     }
 
     abstract execute(input?: string): void;
+
+    protected ignoreWhiteSpace(text: string) {
```
Comment (thilo-behnke) [inline src/main/webapp/app/markdown-editor/commands/command.ts]: - What does this method actually do? If it only removes whitespace in the beginning and end you can use String.prototype.trim, e.g. `text.trim()`.
- A logic heavy method like this needs documentation
- I would avoid to both return a value in a method and change the state of the object (here this.startSpace/this.endspace is set to the identifier of the command). The method name is also misleading here.

I would propose something different:
- startSpace and endSpace should be initialized by the respective command (e.g. italic)
- these identifiers can then be used to check if the selected text is within the command identifiers
- Trimming can be done with a call to `text.trim()`

File: src/main/webapp/app/markdown-editor/commands/italic.command.ts [hunk 8]
```diff
@@ -14,13 +14,15 @@ export class ItalicCommand extends Command {
      */
     execute(): void {
         const selectedText = this.getSelectedText();
+        let textToAdd = '';
 
-        if (selectedText.includes('*')) {
-            const textToAdd = selectedText.slice(1, -1);
+        if (selectedText.charAt(0) === '*' && selectedText.charAt(selectedText.length - 1) === '*') {
```
Comment (thilo-behnke) [inline src/main/webapp/app/markdown-editor/commands/italic.command.ts]: This could be something that is easy to create a method for. `isSelectedTextWithinCommandIdentifiers(text)`.

File: src/main/webapp/app/markdown-editor/commands/command.ts (hunk 7 above)
Comment (HanyaElhashemy) [reply src/main/webapp/app/markdown-editor/commands/command.ts]: Unfortunately `text.trim()` won't do the job here since we need to reinsert the white-spaces after adding the markdowns. So my method counts the white-space characters at the beginning and end of selected text and saves them in other strings (startSapce and endSpace), ignore the white-spaces until markdowns are added around actual text then reinserts the white space.
Example: if the selected text is "I  am  testing" and we select "  am  " using `text.trim()` will return "I**am**testing" which is not the expected behavior.
I agree with you that it is a logic heavy method for what it actually solves.
I noticed that the markdown-editor already deletes more than one successive white-spaces if you click on 'preview' (see screenshot) so another solution would be to use `text.trim()` and only add one white-space per default. In this case we will have about 3 if cases to check where to insert the white-space (at the beginning,end,both or non)
